### DISCUSSION TITLE
Add option to keep the Debug Server always running

### DISCUSSION THREAD

**Discussion by sankasan** [microsoft/debugpy#1268] [Enhancements]
**Summary:**
It would be very helpful if the user could start a debug server and keep it running for the time VSCode keeps running. This allows external processes to initiate the launch of the script while still connecting to the debug server in VSCode.

**Example case:**
Currently I'm working on a C++ API for python. This C++ API ends up in a `.dll` which python scripts load. In Visual Studio 19` I set launch options to call the python interpreter with the script. This allows me to debug the API on the C++ side. If the debug server would keep running on the VSCode side, and let the script connect the the python debugger in VSCode when initiated from Visual Studio, I would be able to hit breakpoints in both C++ (Visual Studio) and Python (VSCode). 

It might be that this feature already exists but I couldn't find it in the documentation.

----
For what its worth, pydevd in Eclipse has this feature, and this keeps me attached to that platform. It is simply too powerful of a development feature.

Below some screenshots of the feature in pydev as exposed in Eclipse.

![pydevd_always_running](https://user-images.githubusercontent.com/2207658/230053084-26710ac2-f0c5-4332-a6ea-01a986b2760b.png)
![pydevd_settings](https://user-images.githubusercontent.com/2207658/230053088-4945bd89-b63c-4e65-976b-36e1cbf72f1d.png)

**Comment by int19h** [comment]
Something like this should be possible if you combine "attach" / "listen" with `"restart": true` in the debug configuration - this causes VSCode (or any other client that supports restarting) to spin up the debug adapter on IDE side again every time the session ends, so effectively you end up with a permanent listener that can be connected to via `debugpy.connect()` as needed.

**Comment by sankasan** [comment]
@int19h I'm not sure I get what you're trying to explain. Specifically the part about `"restart": true`.

Right now I have a python file which at some point has the following:

```
import debugpy
debugpy.listen(5678)
ebugpy.wait_for_client()
```

What do you suggest needs to be done (or perhaps where) to always have the VSCode debug server running so when I launch an interpreter outside of VSCode it will attach to its debugger?

**Comment by int19h** [comment]
On VSCode side, use a debug configuration with `"listen": {"port": ...}` in it to make it wait for an *incoming* connection from the debuggee instead of trying to open an outgoing connection to the same. Then in your Python code, use `debugpy.connect(...)` to make that incoming connection.

By itself, this just reverses the direction of connection. But if you also add `"restart": true` to your debug config, VSCode will restart the session - i.e. will start listening for incoming connections again - once debuggee exits, unless you specifically terminate the session using Stop. This should allow for another connection etc.

**Comment by sankasan** [comment]
@int19h, unfortunately I've not been able to make this work. To simplify testing I created a simple project in CLion and launch the python interpreter from there when starting a debug session. This hits the breakpoints in C++ code when the `.so` file is loaded from python code. However, getting the python script to connect to the VSCode debugger server keeps failing for me. 

I'm probably simply unaware of some fundamentals on how to make VSCode python debugging tick. 

The relevant snippet from the python script:
```
import sys
import debugpy

if __name__ == "__main__":
    # print version information
    print(sys.version)

    # let debugpy connect to the VSCode debug server
    debugpy.connect(5678)
```


This is the `launch.json`:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Remote Attach",
            "type": "python",
            "request": "attach",
            "pathMappings": [
                {
                    "localRoot": "${workspaceFolder}",
                    "remoteRoot": "."
                }
            ],
            "justMyCode": true,
            "listen": {
                "port": 5678,
                "restart": true
            }            
        }
    ]
}
```

I launch VSCode debugger once to make sure the debug server is started. Based on your earlier comments I assume it should keep running afterwards thanks to the `"restart": true` flag. It appears to be running and waiting for a connection.

When I launch the python interpreter from CLion the C++ breakpoint gets hit. It does not hit any python breakpoint in VSCode. VSCode stops waiting for the a client connection, so the retry does not seem to be working. 

It looks like `debugpy` tries to connect but something somewhere fails. I didn't spot anything obvious wrong in the logs. Will continue to dig into it tomorrow but perhaps you see something obvious wrong here already. [debugpy.pydevd.log](https://github.com/microsoft/debugpy/files/11357807/debugpy.pydevd.log)

Sorry to keep this thread open. I really appreciate your you take the time to help me figure this out.

Thanks!

**Comment by sankasan** [comment]
By setting `"justMyCode": false,` I get a bit more info. Does not make much sense to me yet... but something to look into.

![Screenshot_20230429_012924](https://user-images.githubusercontent.com/2207658/235269562-e8d9f527-9767-4108-9e3e-c4aec53f3371.png)

**Comment by int19h** [comment]
To clarify, the "restart" bit needs to be outside of the "listen" block.

With respect to UI for the listening server we're limited by what a debugger extension can do in VSCode. But, basically, so long as it's listening, you should be seeing the debugger toolbar with "Stop" etc - if that goes away, the server is not there anymore.

**Comment by sankasan** [comment]
Yes, I've been able to debug single sessions initiated from CLion and hitting breakpoints in VSCode!

Regarding the `"restart": true` outside of the `listen` block. According to VSCode this is not correct for a configuration: `"Property restart is not allowed."`. When moving it a level higher I get no syntax complains but it doesn't work either. 

Right now the session only works once. Do you happen to have a working sample config?

**Comment by sankasan** [comment]
While certain debuggers seem to support `"restart": true` the python debugger does not appear to do so. The `node` type for example seem to support this. This `launch.json` format seems a bit opaque and clear docs on what is supported for each debugger, even if official, seem to be missing. Or at least I can't find anything more than these:

- https://code.visualstudio.com/Docs/editor/debugging#_launchjson-attributes
- https://code.visualstudio.com/docs/python/debugging#_debugging-specific-app-types

Am I missing something here?

**Comment by int19h** [comment]
There is a bit of a disconnect between what debugpy actually supports, and what the VSCode launch.json schema for its config says it supports. The "restart" bit is very recent, and not fully supported for all types of attach, so it's not in their schema yet. And the squiggles that you get from the editor are from schema validation, so they shouldn't prevent the config from actually being used. 

(One way to make sure that this property is indeed being read by debugpy is to put some nonsense there instead of a boolean - then you should get a message box with an error about it when you actually start debugging.)

**Comment by sankasan** [comment]
Finally got it working, thanks for the support @int19h! 

I was unable to reproduce the `launch.json` errors you described until I moved the python file and .vscode folder outside the C++ (CMake) project folder. Things started working as expected afterwards. Thanks again!
